packages/rum-core/src/domain/foregroundContexts.ts

```diff
@@ -0,0 +1,123 @@
+import {
+  ServerDuration,
+  Configuration,
+  noop,
+  addEventListener,
+  DOM_EVENT,
+  RelativeTime,
+  elapsed,
+  relativeNow,
+  toDuration,
+  toServerDuration,
+} from '@datadog/browser-core'
+import { ForegroundContext, InForegroundPeriod } from '../rawRumEvent.types'
+
+const MAX_NUMBER_OF_FOCUSED_TIME = 500
+
+export interface ForegroundContexts {
+  getInForeground: (startTime: RelativeTime) => ForegroundContext | undefined
+  getInForegroundPeriods: (startTime: RelativeTime, duration: ServerDuration) => InForegroundPeriod[] | undefined
+  stop: () => void
+}
+
+export interface FocusPeriod {
+  start: RelativeTime
+  end?: RelativeTime
+}
+
+let focusPeriods: FocusPeriod[] = []
+
+export function startForegroundContexts(configuration: Configuration) {
+  if (!configuration.isEnabled('track-foreground')) {
+    return {
+      getInForeground: () => undefined,
+      getInForegroundPeriods: () => undefined,
+      stop: noop,
+    }
+  }
+  if (document.hasFocus()) {
+    addNewFocusPeriod()
+  }
+
+  const { stop: stopFocusTracking } = trackFocus(addNewFocusPeriod)
+  const { stop: stopBlurTracking } = trackBlur(closeFocusPeriod)
+  return {
+    getInForeground,
+    getInForegroundPeriods,
+    stop: () => {
+      focusPeriods = []
+      stopFocusTracking()
+      stopBlurTracking()
+    },
+  }
+}
+
+function addNewFocusPeriod() {
+  const now = relativeNow()
+  if (focusPeriods.length > MAX_NUMBER_OF_FOCUSED_TIME) {
+    return
+  }
+  const lastIndex = focusPeriods.length - 1
+  if (lastIndex >= 0 && focusPeriods[lastIndex].end == null) {
+    // already here
+    if (focusPeriods[lastIndex].start === now) {
+      return
+    }
+    // close current periods if we miss a blur event
+    const { start } = focusPeriods[lastIndex]
+    focusPeriods[lastIndex] = {
+      start,
+      end: now,
+    }
+  }
+  focusPeriods.push({
+    start: now,
+  })
```

Comment:
I want to say yes because it means we never closed the previous one. And if we do not it. It will continue to increase. I would like to show that's its two different periods. I first wanted to `end: now - DELAY` but it feels artificial. We can also decide to discard to unclosed period because it's not normal.
Let's add some monitoring because it's not normal

Reply:
>  it means we never closed the previous one. And if we do not it. It will continue to increase.

from a complexity perspective, I'd say let's ignore the new focus and keep the current one open

> I would like to show that's its two different periods.

why? would it be usable somehow?

> We can also decide to discard to unclosed period because it's not normal.

hum, do we prefer to miss focused periods or blur periods?

> Let's add some monitoring because it's not normal

sounds good to me, with the simplest approach in the mean time if we don't have good reason to add extra complexity

Reply:
> hum, do we prefer to miss focused periods or blur periods?

From a replay point of view, I prefer to say to the user that we have a foreground period with nothing happening then saying nothing is happing while in reality there are stuff happening.

So I would prefer to miss blur periods.
I am merging the two periods